Run unit tests with truffle-compiler

Akirathan opened the pull request:
Fixes #8374

### Pull Request Description

`runtime/test` and `language-server/test` runs with truffle-compiler module. This should speed up these tests. Moreover, it fixes the `withDebug` command.

### Important Notes

- Add `runtime-fat-jar` sbt project.
  - A *meta* project that just contains the logic for assembling `runtime.jar` fat jar.
  - Increases readability.
  - No need to copy any classes from other `runtime-*` projects into `runtime-with-instruments/target` directory.
- Add `runtime-test-instruments` project.
  - This project is an *explicit JPMS module*, i.e., it has `module-info.java`.
  - Its sole purpose is to provide truffle instruments that are used in `runtime/test`.
  - It uses reflection to access classes from `runtime`.
- Merge `runtime-with-instruments` and `runtime-with-polyglot` into `runtime/Test`. Those projects contained only tests.
- Add a caching mechanism so that classes from other `runtime-*` projects are not copied every time into `runtime-fat-jar/target`.
- `runtime/test` are run with patched `org.enso.runtime` module such that all the instruments (and test classes) are included in the module. Moreover, the `--patch-module` is required to avoid reassembling `runtime.jar` fat jar before every test.
- Various sbt refactorings.
  - Add `JPMSPlugin` sbt auto plugin that makes handling of the JPMS modules related stuff easier.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

Commit:
logging-service-logback is JPMS module

Commit:
runtime uses truffle-compiler in tests

Commit:
Revert "logging-service-logback is JPMS module"

Commit:
Refactor Test/options to a separate task

Commit:
Provide a default logback-test logging config for tests

Commit:
Do not add runtime-with-instruments to --patch-module

Commit:
Refactor mutable ListBuffer to List

hubertp commented on build.sbt:
```diff
@@ -519,6 +522,105 @@ lazy val componentModulesPaths =
+    runtimeMod.map(_.getAbsolutePath) ++
+    graalMods.map(_.data.getAbsolutePath) ++
+    loggingMods.map(_.getAbsolutePath)
+  // We can't use org.enso.logger.TestLogProvider (or anything from our own logging framework here) because it is not
```
That would mean it will render our configuration useless?

Akirathan replied:
Useless is a very harsh word in general. It just cannot be used for testing, if we want to start the forked JVM with `--module-path`. The problem is [split packages](https://stackoverflow.com/a/51828213/4816269), i.e., we have for example `org.enso.logging` package distributed in many projects, and if we would like to produce a single Jar module per single project, we cannot succeed.

Do you think it is problematic to use just plain `ch.qos.logback` for the test logging?

hubertp replied:
> Do you think it is problematic to use just plain ch.qos.logback for the test logging?

Well, I think _useless_ is the right word in this case because there is no point of `TestLogProvider` anymore, unless I'm missing something. I'm not sure the setup will work now as expected, though. Can you double check that the output from `sbt> test` is now comparable?

Akirathan replied:
At the end, I am using `TestLogProvider` - a342562c2dd98b2ab5c1bd22e1ccb8ba82f82582. So there should be no difference in the test output with develop.

Commit:
Introduce runtime-fat-jar project.

This project bundles all the functionality for assembling the `runtime.jar` fat jar.

Commit:
compileModuleInfo task does not copy anything unless necessary

Commit:
Merge branch 'develop' into wip/akirathan/8374-compiler-to-tests

# Conflicts:
#	build.sbt

Commit:
JPMSUtils.filterModules filters only distinct modules

Commit:
Make sure that all GraalVM modules (including truffle langs) are in --module-path in runtime/test

Commit:
Create new module org.enso.interpreter.test

This module will contain all the truffle instruments required for testing

Commit:
module-info.java is not compiled every time

Commit:
Revert "Create new module org.enso.interpreter.test"

This reverts commit 4a9c9629c68bbc58e000183d7683831b1d664770.

Commit:
Add skeleton of JPMSPlugin auto plugin.

Commit:
Add runtime-test-instruments project.

Commit:
Fix VectorTests in runtime

Commit:
Fix the rest of runtime tests

Commit:
runtime-with-instruments/test run with truffle-compiler

Commit:
fmt

Commit:
Fix some sbt issues with lazy val settings

Akirathan commented:
So far, `runtime/test` and `runtime-with-instruments/test` work, and they both run with `truffle-compiler`. Moreover, they do not assemble any fat jars, which should have a positive impact on build times.

JaroslavTulach commented on engine/runtime-test-instruments/src/main/java/module-info.java:
```diff
@@ -0,0 +1,9 @@
+
+  exports org.enso.interpreter.test.instruments;
+  
+  provides com.oracle.truffle.api.instrumentation.provider.TruffleInstrumentProvider with
```
Separate module & module info for test only instruments is a good idea.

JaroslavTulach commented on engine/runtime-test-instruments/src/main/java/org/enso/interpreter/test/instruments/CodeIdsTestInstrument.java:
```diff
@@ -21,8 +21,24 @@
     services = CodeIdsTestInstrument.class)
 public class CodeIdsTestInstrument extends TruffleInstrument {
```
All you need to do here is to register the instruments, so Truffle finds them in a separate module.

JaroslavTulach commented on engine/runtime-test-instruments/src/main/java/org/enso/interpreter/test/instruments/CodeIdsTestInstrument.java:
```diff
@@ -21,8 +21,24 @@
+  static {
+    try {
+      exprNodeClass = Class.forName("org.enso.interpreter.node.ExpressionNode");
+      exprNodeGetId = exprNodeClass.getMethod("getId");
```
Create a service class:
```java
interface GetId {
   UUID getId(Node n);
}
```
and `require service` it in `module info`. Obtain its implementation via `ServiceLoader`. Register its implementation in `runtime-xyz/test` via `@ServiceProvider` annotation.

I can't guarantee it is going to work 100% this way, but it may work fine (btw. use `org.openide.util.Lookup` instead of `ServiceLoader`) if you hit problems between modules/unnamed modules.

JaroslavTulach replied:
The `GetId` interface may be found by [Lookups.metaInfServices](https://github.com/enso-org/enso/blob/857e87430d59a01ef30017a88b892940f73964e9/lib/java/persistance/src/main/java/org/enso/persist/PerMap.java#L18).

Akirathan replied:
Done in b98526ee882e4c10ce56d90b6750fd6e3b1754a1

JaroslavTulach reviewed (commented):
Rather than using reflection: invert the dependencies.

Commit:
Use service provider instead of reflection in runtime-test-instruments

Commit:
Port NodeCountingTestInstrument into runtime-test-instruments

Commit:
Merge runtime-with-instruments project into runtime/Test

Commit:
Merge runtime-with-polyglot project into runtime/Test

Commit:
Run runtime/test without -ea

Commit:
fmt

Commit:
Merge branch 'develop' into wip/akirathan/8374-compiler-to-tests

Commit:
Include insight tool in runtime/Test

Commit:
withDebug command has better error message

Commit:
Remove BigNumberTest.averageOfMixedArrayOverDouble.

The test is now failing on "Cannot convert big integer to double", which is now expected.

Commit:
Fix context dispose in MetaObjectTest

Commit:
project-manager/test runs with truffle-compiler

Commit:
language-server/test runs with truffle-compiler

Commit:
JPMSUtils.compileModuleInfo takes an extraModulePath optional argument

Commit:
Merge branch 'develop' into wip/akirathan/8374-compiler-to-tests

Commit:
Small fixes after merge

Commit:
langauge-server/test uses enso TestLogProvider

Commit:
Add necessary modules on modulePath to language-server/test

Commit:
Add --patch-module option to language-server/test

Commit:
runtime/test uses Enso TestLogProvider

Commit:
Fix DebuggingEnsoTest.

Stepping over / into behavior is reverted to before 5a7ad6bfe4fe6c0a945e1afb7c3ec3e58b937293

Commit:
Make sure runtime/test is run with -ea

Commit:
Make sure runtime/test is run with -ea

Commit:
Remove dependency on com.sandinh.sbt-java-module-info plugin

Commit:
CLeanup some unused methods

Commit:
fmt

Commit:
project-manager/test uses TestLogProvider

Commit:
Remove unused logback-test.xml resource

JaroslavTulach reviewed (approved):
Let's get this in!
